web-api/src/persistence/postgres/utils/transactions.ts

```diff
@@ -0,0 +1,55 @@
+import { settlePromises } from '@web-api/utilities/settlePromises';
+import {
+  ConnectionInfo,
+  ConnectionStore,
+  getDb,
+} from '@web-api/persistence/postgres/databaseConnection';
+
+export async function withTransaction<T>(fn: () => Promise<T>): Promise<T> {
+  // If we're already in a transaction, just run the callback directly.
+  if (inTransaction()) {
+    return fn();
+  }
+
+  const db = await getDb();
+
+  // We'll capture the store here so it's available after commit.
+  let transactionStore: ConnectionInfo = {} as ConnectionInfo;
+
+  // Start the transaction; Kysely handles BEGIN/COMMIT/ROLLBACK.
+  const result = await db.transaction().execute(async trx => {
```

Comment:
Kysely guarantees that only one connection is used for the transaction, and that only one transaction is on that connection at a time,.